x/sync/network_server.go

```diff
@@ -232,48 +250,77 @@ func (s *NetworkServer) HandleRangeProofRequest(
 		return nil // dropping request
 	}
 
-	// override limit if it is greater than maxKeyValuesLimit
-	keyLimit := req.KeyLimit
-	if keyLimit > maxKeyValuesLimit {
-		keyLimit = maxKeyValuesLimit
+	// override limits if they exceed caps
+	if req.KeyLimit > maxKeyValuesLimit {
+		req.KeyLimit = maxKeyValuesLimit
 	}
-	bytesLimit := int(req.BytesLimit)
-	if bytesLimit > maxByteSizeLimit {
-		bytesLimit = maxByteSizeLimit
+
+	if req.BytesLimit > maxByteSizeLimit {
+		req.BytesLimit = maxByteSizeLimit
+	}
+
+	proofBytes, err := getRangeProof(
+		ctx,
+		s.db,
+		req,
+		func(rangeProof *merkledb.RangeProof) ([]byte, error) {
+			return proto.Marshal(rangeProof.ToProto())
+		},
+	)
+	if err != nil {
+		return err
 	}
+	// TODO handle this fatal error
+	return s.appSender.SendAppResponse(ctx, nodeID, requestID, proofBytes)
+}
+
+// Get the range proof specified by [req].
+// If the generated proof is too large, the key limit is reduced
+// and the proof is regenerated. This process is repeated until
+// the proof is smaller than [req.BytesLimit].
+// When a sufficiently small proof is generated, returns it.
+// If no sufficiently small proof can be generated, returns [ErrMinProofSizeIsTooLarge].
+func getRangeProof(
+	ctx context.Context,
+	db DB,
+	req *pb.SyncGetRangeProofRequest,
+	marshalFunc func(*merkledb.RangeProof) ([]byte, error),
+) ([]byte, error) {
+	root, err := ids.ToID(req.RootHash)
+	if err != nil {
+		return nil, err
+	}
+
+	keyLimit := int(req.KeyLimit)
+
 	for keyLimit > 0 {
-		root, err := ids.ToID(req.RootHash)
-		if err != nil {
-			return err
-		}
-		rangeProof, err := s.db.GetRangeProofAtRoot(ctx, root, req.StartKey, req.EndKey, int(keyLimit))
+		rangeProof, err := db.GetRangeProofAtRoot(
+			ctx,
+			root,
+			req.StartKey,
+			req.EndKey,
+			keyLimit,
+		)
 		if err != nil {
-			// handle expected errors so clients cannot cause servers to spam warning logs.
-			if errors.Is(err, merkledb.ErrRootIDNotPresent) {
-				s.log.Debug(
-					"dropping invalid range proof request",
-					zap.Stringer("nodeID", nodeID),
-					zap.Uint32("requestID", requestID),
-					zap.Stringer("req", req),
-					zap.Error(err),
-				)
-				return nil // dropping request
+			if errors.Is(err, merkledb.ErrInsufficientHistory) {
+				return nil, nil // drop request
```

Comment:
I could add a log here but it seems kind of annoying. I'd also be open to checking for `ErrInsufficientHistory` in the caller.